python/shotgun_model/shotgun_entity_model.py

```diff
@@ -0,0 +1,249 @@
+        """
+        self._entity_icons = {}
+        self._step_swatch_icons = {}
+
+        # make sure fields is valid:
+        fields = fields or []
+        # for backwards compatibility, make sure certain fields are added:
+        fields = list(set(fields + ["image", "sg_status_list", "description"]))
+
+        ## folder icon
+        self._default_icon = QtGui.QIcon(QtGui.QPixmap(":/tk-framework-shotgunutils/icon_Folder.png"))
+
+        ShotgunModel.__init__(self, 
+                              parent = parent,
+                              download_thumbs = download_thumbs,
+                              schema_generation = schema_generation,
+                              bg_load_thumbs = bg_load_thumbs,
+                              bg_task_manager = bg_task_manager)
+
+        # load the data from the cache:
+        self._load_data(entity_type, filters, hierarchy, fields)
+
+    def destroy(self):
+        """
+        Call to clean-up the model when it is finished with
+        """
+        ShotgunModel.destroy(self)
+        self._entity_icons = {}
+        self._step_swatch_icons = {}
+        self._default_icon = None
+
+    def get_entity_icon(self, entity_type):
+        """
+        Retrieve the icon for the specified entity type if available.
+
+        :param entity_type: The entity type to retrieve the icon for
+        :returns:           A QIcon if an icon was found for the specified entity
+                            type, otherwise None.
+        """
+        icon = None
+        if entity_type in self._entity_icons:
+            # we've previously asked for the icon
+            icon = self._entity_icons[entity_type]
+        else:
+            # see if we have the icon in the resources:
+            icon_path = ":/tk-framework-shotgunutils/icon_%s_dark.png" % entity_type
+            if QtCore.QFile.exists(icon_path): 
+                # create the new icon from this resource:
+                icon = QtGui.QIcon(QtGui.QPixmap(icon_path))
+            self._entity_icons[entity_type] = icon
+
+        return QtGui.QIcon(icon) if icon else None
+
+    def get_entities(self, item):
+        """
+        Get entities for the current item by traversing up the tree and pulling 
+        entity information from each item if possible
+
+        :param item:    The item to find entities for
+        :returns:       A list of entities in the order they were found starting
+                        from the specified item.
+        """
+        current_item = item
+        entities = []
+        while current_item:
+            item_entity = self.get_entity(current_item)
+            if item_entity:
+                entities.append(item_entity)
+            current_item = current_item.parent()
+
+        return entities
+
+    def get_entity(self, item):
+        """
+        Get the Shotgun entity details for the specified model item.
```

Comment:
Again, I think an example here would be helpful to illustrate what this method does that `item.get_sg_data()` doesn't!

Also, maybe it's weird that it returns different things. This is something I sometimes tend to avoid in my code.

So if you call this on a intermediate node for a model for `Shot`s with hierarchy `[project, sg_sequence, code]` you would get back `{type: Sequence, id: 1234, name: foo}` but if you call it for a model for `Sequence`s with hierarchy `[project, code]` (e..g sequence is a leaf node) you would potentially get more fields - and perhaps more importantly `code` rather than `name`. I wonder if this should always return name, type and id for all nodes regardless if they are leaves or not. Would that make it easier to use in the calling code? or maybe even just type and id? (but then again it's often useful to have a name)